lasy/profiles/speckle_profile.py

```diff
@@ -0,0 +1,371 @@
+            elif "ISI" in self.lsType.upper():
+                complex_amp = np.stack(
+                    [
+                        np.stack(
+                            [
+                                gen_gaussian_time_series(
+                                    stochastic_process_time.size,
+                                    2 * self.laser_bandwidth,
+                                    1,
+                                )
+                                for _i in range(self.n_beamlets[1])
+                            ]
+                        )
+                        for _j in range(self.n_beamlets[0])
+                    ]
+                )
+                return stochastic_process_time, complex_amp
+
+        if "GP" in self.lsType.upper():
+            time_ext, timeSeries = init_GS_timeseries()
+        else:
+            time_ext, timeSeries = stochastic_process_time, None
+
+        def beamlets_complex_amplitude(t_now, lsType="FM SSD"):
+            if lsType.upper() == "FM SSD":
+                phase_t = self.phase_mod_amp[0] * np.sin(
+                    phase_mod_phase[0]
+                    + 2
+                    * np.pi
+                    * phase_mod_freq[0]
+                    * (t_now - X_lens_matrix * td[0] / self.n_beamlets[0])
+                ) + self.phase_mod_amp[1] * np.sin(
+                    phase_mod_phase[1]
+                    + 2
+                    * np.pi
+                    * phase_mod_freq[1]
+                    * (t_now - Y_lens_matrix * td[1] / self.n_beamlets[1])
+                )
+                return np.exp(1j * phase_t)
+            elif lsType.upper() == "GP RPM SSD":
+                phase_t = np.interp(
+                    t_now + X_lens_index_matrix * td[0] / self.n_beamlets[0],
+                    time_ext,
+                    timeSeries[0],
+                ) + np.interp(
+                    t_now + Y_lens_index_matrix * td[1] / self.n_beamlets[1],
+                    time_ext,
+                    timeSeries[1],
+                )
+                return np.exp(1j * phase_t)
+            elif lsType.upper() == "GP ISI":
+                return timeSeries[:, :, int(round(t_now / self.tu))]
+            else:
+                raise NotImplementedError
+
+        exp_phase_plate = np.exp(1j * phase_plate)
+        lambda_fnum = self.wavelength * self.focal_length / self.beam_aperture
+        X_focus_matrix = x[:, :, 0] / lambda_fnum[0]
+        Y_focus_matrix = y[:, :, 0] / lambda_fnum[1]
+        x_focus_list = X_focus_matrix[:, 0]
+        y_focus_list = Y_focus_matrix[0, :]
+        x_phase_matrix = np.exp(
+            -2
+            * np.pi
+            * 1j
+            / self.n_beamlets[0]
+            * np.einsum("i,j", x_lens_list, x_focus_list)
+        )
+        y_phase_matrix = np.exp(
+            -2
+            * np.pi
+            * 1j
+            / self.n_beamlets[1]
+            * np.einsum("i,j", y_lens_list, y_focus_list)
+        )
```

Comment:
Maybe a more readable way to write this would be:
```suggestion
        # Position of the beamlet centers in the near field
        x_lens = self.beam_aperture[0]/self.n_beamlets[0] * np.linspace(
            -0.5 * (self.n_beamlets[0] - 1),
            0.5 * (self.n_beamlets[0] - 1),
            num=self.n_beamlets[0],
        )
        # Corresponding k vector for each beamlet
        k_x = 2*np.pi/self.wavelength * x_lens / self.focal_length
        # Corresponding phase term for each beamlet at each point in the focal plane
        x_focus_list = x[:,0,0]
        x_phase_matrix = np.exp( -1.j * k_x[:, np.newaxis] * x_focus_list[np.newaxis, :]
```
with the same thing for `y`

Reply:
Also: `k_x` and `k_y` could be potentially calculated in the constructor of this class.
(Since they do not require knowledge of the size of `x` and `y`)